src/Serokell/Util/Group.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE ExplicitForAll      #-}
+{-# LANGUAGE NoImplicitPrelude   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+
+-- | This module introduces functions for grouping.
+
+module Serokell.Util.Group
+       ( groupBy
+       , groupSingleBy
+       ) where
+
+import Universum
+
+import Data.List.NonEmpty ((<|))
+
+import qualified Data.HashMap.Strict as HM
+
+{- | Groups elements using results of the given function as keys.
+
+>>> groupBy even [1,2,3,4,5,6]
+fromList [(False,5 :| [3,1]),(True,6 :| [4,2])]
+
+-}
+groupBy :: forall t a b . (Container t, Element t ~ a, Eq b, Hashable b)
+        => (a -> b) -> t -> HashMap b (NonEmpty a)
+groupBy f = foldl' (hmGroup f) mempty
+  where
+    hmGroup :: (a -> b) -> a -> HashMap b (NonEmpty a) -> HashMap b (NonEmpty a)
+    hmGroup fun x hm =
+        let key = fun x
+            val = case HM.lookup key hm of
+                Nothing -> one x
+                Just xs -> x <| xs
+        in HM.insert key val hm
+
+{- | Similar to 'groupBy' but keeps only one element as value.
+
+>>> groupSingleBy even [1,2,3,4,5,6]
+fromList [(False,1),(True,2)]
+
+-}
+groupSingleBy :: forall t a b . (Container t, Element t ~ a, Eq b, Hashable b)
+              => (a -> b) -> t -> HashMap b a
+groupSingleBy f = foldl' (hmGroup f) mempty
```

Comment:
I think you can name this function as `groupOneBy` to make it shorter. Though, naming might not be perfect because this is not _grouping_ actually... Maybe `mapBy` is better?..